guides/Postgres Query Optimization + Diagnostic Tooling.md

```diff
@@ -0,0 +1,79 @@
+---
+Description: PostgreSQL is a database management system that utilizes SQL and includes features like foreign keys, updatable views, and various query optimization tools. This article will outline some of the query optimization techniques in Postgres using diagnostic tools.
+---
+# Postgres Query Optimization + Diagnostic Tooling
+
+## Scope
+Query optimization involves identifying the most efficient way to execute a query, considering the structure of the database and available indexes. The Postgres query optimizer can perform this task. Proper optimization techniques can help reduce costs and enhance environmental sustainability by using resources more efficiently.
+
+## Postgres Query Optimizer
+The PostgreSQL Optimizer generates a plan for each query it receives, and the EXPLAIN command allows one to view these plans. The EXPLAIN ANALYZE command can be used to evaluate the accuracy of these plans.  
+
+ For example:
+ 
+   ```EXPLAIN ANALYZE SELECT * FROM customer;```
+ 
+ The execution plan of this query is:
+ 
+  ![](/.gitbook/assets/query-optimization-images/picture1.PNG)
+
+## Postgres Query Optimization and Performance Tuning Methods
+
+In this section, you will learn a few optimization techniques along with examples and query execution plans.
+
+### Use of Indexes
+An index serves as an additional access structure that enables the rapid location and retrieval of specific data. In the absence of an index, a query would need to search the entire table to find the desired information. By indexing columns commonly used in WHERE or JOIN clauses, PostgreSQL can swiftly locate the desired data.
+
+For instance, the following picture depicts the query execution plan without any indexes.
+
+ ![](/.gitbook/assets/query-optimization-images/picture2.PNG)
+
+The execution plan displayed in this image depicts the impact of adding an index on the address_id column. The addition of the index has resulted in a noticeable improvement in both Planning Time and Execution Time.
+
+ ![](/.gitbook/assets/query-optimization-images/picture3.PNG)
+
+### Optimize SELECT Clause
+
+Efficiency issues in queries can be addressed by using specific column names instead of the *  argument in SELECT statements. Additionally, it is important to execute the HAVING clause after filtering the data with the SELECT statement, as SELECT acts as a filter. To illustrate, replacing * with Payment_id in Group by can improve the query performance. You can see the change in execution plans in both the queries.
+
+![](/.gitbook/assets/query-optimization-images/picture4.PNG)
+
+
+![](/.gitbook/assets/query-optimization-images/picture5.PNG)
+
+The performance difference is not significant, but according to the sources reviewed in the literature study, modifying the use of * is expected to enhance performance.
```

Comment:
```suggestion
While the performance difference is not significant in this example, avoiding use of `*` can often greatly enhance performance when using columnar tables.
```

Good opportunity here to mention that not using `*` can really help columnar queries.